Second week

Sinjvf opened the pull request:
Задание 2: Добавьте в приложение экран с текущей погодой (без прогноза). Обновление должно работать не только вручную, но и по расписанию (не забудьте добавить интервал обновления в настройки!). Выбор города пока не требуется (пусть будет Москва). API погоды и набор иконок - на ваше усмотрение.

Commit:
Add:
 - ViewModel для базовой активити. Отслеживание смены фрагмента перенесено в этот ViewModel. Теперь при повороте мы остаемся на том же фрагменте, что и были

 Change:
 - Для выставления титульного текста из фрагмента кастуем getActivity()  к интерфейсу, а не к конкретному типу активити

Sinjvf committed:
Intermediate commit. Fix last homework:
 - add release build
 - separate icons and full package names on debug and release versions
 - add proguard rules for libraries

Sinjvf committed:
Intermediate commit. Fix last homework:
 - add nonnull annotations
 - add comments for classes
 - add test for testing toolbar title
 - revert first changing fragment logic: if OpenNewFragmentEvent's field addToBackStack is true we add it to backStack
 - remove eventBus: we probably don't need a complex moving between fragments
 - fix proguard rules fot lifecycle.
 - Butterknife  unbinder init  move to BaseFragments class
 - logic move from OnCreateView to OnViewCreated in fragments
 - formatted string instead of concat
 - delete unused tests

Sinjvf committed:
second homework:

 - add classes for retrofit behaviour
 - add data classes witch are used in responses
 - print current temperature in max\min for day in Main Fragment
 - add Dialogs class for showing dialogs
 - add common response handler - show error dialogs when request failed

Sinjvf committed:
second homework:

 - add autoUpdating data when main screen is resumed;
 - add select  updating interval in settings
 - add: show last updating time

Sinjvf committed:
second homework:

 - del unused icons
 - add weather icon(load with Picasso), humidity and wind speed
 - swipe refresh in  the weather screen
 - proguard rules for Picasso

Commit:
fix bugs:
 - default update interval from shared prefs returns 10 minutes instead of 0
 - stop show refreshing after response with error
 - base fragment has method to processing progress and base response handler send events to it

shmakova commented on app/build.gradle:
```diff
@@ -30,7 +43,7 @@ android {
 
 ext {
-    supportVersion = '26+'
+    supportVers = '26+'
```
Нужно указывать номер версии без + для ускорения сборки

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/activity/MainActivity.java:
```diff
@@ -1,65 +1,65 @@
 
-public class MainActivity extends AppCompatActivity
-        implements NavigationView.OnNavigationItemSelectedListener {
+    protected final String TAG = "tag:" + getClass().getSimpleName();
```
Не используешь нигде. Для логов удобно использовать Timber, там например не надо постоянно TAG прописывать

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/activity/MainActivity.java:
```diff
@@ -93,33 +92,27 @@ public boolean onNavigationItemSelected(MenuItem item) {
+    public void changeFragment(@NonNull OpenNewFragmentEvent event) {
+        FragmentManager fm = getSupportFragmentManager();
+        if (event.isAddToBackStack())
+            fm.beginTransaction().addToBackStack(null).replace(R.id.fragment_container, event.getFgm()).commit();
```
Нужно в скобочки оборачивать, даже если один statement

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/activity/MainActivityViewModel.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @Override
+    protected void onCleared() {
```
Методы, которые вызывают только super, можно убирать из кода

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/base_util/BaseFragment.java:
```diff
@@ -4,37 +4,57 @@
-            ((MainActivity) getActivity()).initToolbar(text);
+            ((MainActivityInterface) getActivity()).setTitleText(titleResId);
         } catch (ClassCastException e) {
             e.printStackTrace();
```
Avoid printStackTrace(); use a logger call instead.

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/base_util/BaseFragmentViewModel.java:
```diff
@@ -4,33 +4,91 @@
+    protected SingleObserver<Integer> progressObserver(){
+        return new SingleObserver<Integer>() {
+            @Override
+            public void onSubscribe(@io.reactivex.annotations.NonNull Disposable d) {}
```
А обычные android.support.annotation.NonNull тут нельзя использовать?

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/base_util/BaseFragmentViewModel.java:
```diff
@@ -4,33 +4,91 @@
+    }
+
+    public MutableLiveData<Integer> getProgress() {
+        if (progres == null){
```
Опечатка в progress

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/fragments/about/AboutFragment.java:
```diff
@@ -39,14 +40,17 @@ public void onActivityCreated(@Nullable Bundle savedInstanceState) {
-        return rootView;
     }
 
     private void setVers() {
```
Непонятно почему экономишь буквы

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/fragments/about/AboutFragment.java:
```diff
@@ -39,14 +40,17 @@ public void onActivityCreated(@Nullable Bundle savedInstanceState) {
 
     private void setVers() {
-        String name = getString(R.string.about_version) + " " + BuildConfig.VERSION_NAME;
+        String name = String.format(getString(R.string.about_version), BuildConfig.VERSION_NAME);
```
String name = getString(R.string.about_version, BuildConfig.VERSION_NAME);

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/fragments/about/AboutViewModel.java:
```diff
@@ -1,21 +1,22 @@
     @Override
-    protected String getTitleText() {
-        return getApplication().getResources().getString(R.string.menu_about);
+    protected int getTitleStringId() {
```
Можно добавить аннотацию `@StringRes`

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/retrofit/ServiceContract.java:
```diff
@@ -0,0 +1,62 @@
+ * Describes server configuration
+ */
+
+class ServiceContract {
```
Можно попробовать использовать даггер для инициализации ретрофита, okhttp, gson, интерсепторов и т.п.

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/retrofit/ServiceContract.java:
```diff
@@ -0,0 +1,62 @@
+
+    interface WeatherAPI {
+        @GET(WEATHER)
+        Call<WeatherResponse> getWeather(@QueryMap Map<String, String> map);
```
Можно добавить вот это при инициализации ретрофита: `.addCallAdapterFactory(RxJava2CallAdapterFactory.create())` и возвращать Single вместо Call

Sinjvf replied:
Да, я как раз планировала получше разобраться с Rx и использовать его. Как раз для этого в ServiceHandler'е метод callService, чтобы потом переписать только его, а не каждое из обращений к серверу.

shmakova commented on app/src/androidTest/java/ru/mobilization/sinjvf/yamblzweather/CheckToolbarTest.java:
```diff
@@ -0,0 +1,92 @@
+
+
+@RunWith(AndroidJUnit4.class)
+public class CheckToolbarTest {
```
Тест 👍

shmakova commented on app/src/androidTest/java/ru/mobilization/sinjvf/yamblzweather/CheckToolbarTest.java:
```diff
@@ -0,0 +1,92 @@
+
+import static android.support.test.espresso.Espresso.onView;
+import static android.support.test.espresso.action.ViewActions.click;
+import static android.support.test.espresso.action.ViewActions.swipeRight;
```
Тут есть unused imports, их надо убирать (при коммите в студии можно поставить галочку Optimize imports)

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/base_util/WeatherApplication.java:
```diff
@@ -0,0 +1,19 @@
+    @Override
+    public void onCreate() {
+        super.onCreate();
+        ServiceHandler.getInstance(getApplicationContext());
```
Лучше через даггер всё это делать

Sinjvf replied:
Хорошо, посмотрю, как это делается

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/ui/SelectIntervalDialog.java:
```diff
@@ -0,0 +1,117 @@
+ * Class for showing dialogs to user
+ */
+
+public class SelectIntervalDialog extends DialogFragment {
```
Лучше оставлять в названии класса Fragment для читабельности

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/ui/SelectIntervalDialog.java:
```diff
@@ -0,0 +1,117 @@
+        radio30.setText(String.format(getString(R.string.n_min), 30));
+        radio60.setText(String.format(getString(R.string.n_min), 60));
+        switch ((int)(Preferenses.getIntervalTime(getContext())/Utils.MINUTE)){
+            case 15:
```
Все магические числа нужно вынести в константы

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/ui/SelectIntervalDialog.java:
```diff
@@ -0,0 +1,117 @@
+                break;
+        }
+        if (action != null) {
+            Single.fromObservable(Observable.fromArray(time)).subscribe(action);
```
Не отписываешься

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/ui/Dialogs.java:
```diff
@@ -0,0 +1,113 @@
+    @BindView(R.id.cancel)
+    TextView cancel;
+
+    Unbinder unbinder;
```
Можно вынести в базовый класс

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/utils/Utils.java:
```diff
@@ -0,0 +1,47 @@
+ */
+
+public class Utils {
+    public static long MINUTE = 1000 * 60;
```
public static final

Sinjvf replied:
ок

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/ui/SelectIntervalDialog.java:
```diff
@@ -0,0 +1,117 @@
+        int id = radioGroup.getCheckedRadioButtonId();
+        switch (id){
+            case R.id.radio_10:
+                time = 10* Utils.MINUTE;
```
TimeUnit.MINUTES.toMillis(10)

shmakova commented on app/src/main/java/ru/mobilization/sinjvf/yamblzweather/retrofit/BaseResponseCallback.java:
```diff
@@ -0,0 +1,69 @@
+ * Base handler
+ */
+
+public  class BaseResponseCallback<T> implements Callback<T> {
```
Непонятно, почему этот класс лежит в этом пакете (вроде как тут просто должна быть работа с сетью и данными, но тут еще и фрагментами идет работа)

Sinjvf replied:
Он там, тк обрабатывает ответы сервера. Но, возможно, действительно это не очень логично и стоит перенести его в base_util

shmakova reviewed (commented):
Плюс за Android Architecture Components, за эспрессо тест.
Нужно обратить внимание на codestyle, можно взять в помощь статических анализаторов отсюда https://github.com/artem-zinnatullin/qualitymatters
Очень запутанная архитектура и не очень удачное разделение по пакетам (например во fragments у тебя лежат viewmodel классы)
Слой model размазан по всему приложению

Sinjvf commented:
Мне кажется более логичным складывать вместе компоненты отвечающие за один экран. Да, слой model "размазывается", но так лучше видно с какой моделью работает каждый фрагмент. Пожалуй, название пакета "fragments" действительно неудачное, поищу более подходящее.

shmakova commented:
Да, это наиболее логично, можно просто убрать пакет fragments и разделять по фичам

Sinjvf committed:
fix last homework:

 - rename fragment package
 - Timber instead of logs
 - del eventBus from gradle and proguard rules
 - all magic numbers - to constant
 - no printStackTrace in catch sections
 - word "fragment" in in fragment dialog classes names

Commit:
changes:
 - retrofit with Rx
 - main response handler - in baseFragmentViewModel

Commit:
add:
 - missing proguard annotations rule

shmakova commented:
Можешь объяснить, как у тебя работает обновление по расписанию?

Sinjvf commented:
Могу. Поскольку у нас есть только одно место, где погода выводится, и нет кеширования результатов, то я решила, что пока нет необходимости в отдельном сервисе или AlarmManager'e\JobShceduler'e для обновления результатов.

Если открыт экран с погодой, то раз в n минут(n задается в настройках) идет запрос на сервер и происходит обновление экрана. "Раз в n минут" - handler с отложенным post запросом
